packages/Kokkos/src/DTK_PointCloud_def.hpp

```diff
@@ -0,0 +1,235 @@
+ * \brief Point cloud.
+ */
+//---------------------------------------------------------------------------//
+
+#ifndef DTK_POINTCLOUD_DEF_HPP
+#define DTK_POINTCLOUD_DEF_HPP
+
+#include <DTK_DBC.hpp>
+#include <DTK_KokkosHelpers.hpp>
+
+#include <Teuchos_CommHelpers.hpp>
+#include <Teuchos_Ptr.hpp>
+
+#include <Kokkos_Core.hpp>
+
+#include <limits>
+#include <type_traits>
+
+namespace DataTransferKit
+{
+//---------------------------------------------------------------------------//
+// Local bounding box Kokkos functor.
+//---------------------------------------------------------------------------//
+//! Functor to compute the local bounding box of the point cloud.
+template <class SC, class LO, class GO, class NO>
+class LocalBoundingBoxFunctor
+{
+  public:
+    //! Coordinate type alias.
+    using coordinate_view =
+        typename PointCloud<SC, LO, GO, NO>::coordinate_view;
+    typedef SC value_type[];
+    using size_type = typename coordinate_view::traits::size_type;
+
+    //! Constructor.
+    LocalBoundingBoxFunctor( const coordinate_view coordinates )
+        : _coords( coordinates )
+        , value_count( 6 )
+    { /* ... */
+    }
+
+    //! Reduction operator.
+    KOKKOS_INLINE_FUNCTION
+    void operator()( const size_type p, value_type update ) const
+    {
+        for ( size_t d = 0; d < _coords.extent( 1 ); ++d )
+        {
+            update[d] = KokkosHelpers::min( _coords( p, d ), update[d] );
+            update[d + 3] =
+                KokkosHelpers::max( _coords( p, d ), update[d + 3] );
+        }
+    }
+
+    //! Join operator.
+    KOKKOS_INLINE_FUNCTION
+    void join( volatile value_type dst, const volatile value_type src ) const
+    {
+        for ( size_t d = 0; d < _coords.extent( 1 ); ++d )
+        {
+            dst[d] = KokkosHelpers::min( src[d], dst[d] );
+            dst[d + 3] = KokkosHelpers::max( src[d + 3], dst[d + 3] );
+        }
+    }
+
+    //! Init operator
+    KOKKOS_INLINE_FUNCTION
+    void init( value_type dst ) const
+    {
+        for ( size_t d = 0; d < 3; ++d )
+        {
+            dst[d] = std::numeric_limits<SC>::max();
+            dst[d + 3] = -std::numeric_limits<SC>::max();
+        }
+    }
+
+  public:
+    // Point cloud coordinates.
+    coordinate_view _coords;
+
+    // Box size - specifically named this way for Kokkos
+    size_type value_count;
+};
+
+//---------------------------------------------------------------------------//
+// Class Implementation.
+//---------------------------------------------------------------------------//
+/*!
+  \brief Constructor.
+*/
+template <class SC, class LO, class GO, class NO>
+PointCloud<SC, LO, GO, NO>::PointCloud(
+    const Teuchos::RCP<const Teuchos::Comm<int>> &comm,
+    const global_id_view global_ids, const coordinate_view coordinates )
+    : _comm( comm )
+    , _global_ids( global_ids )
+    , _coordinates( coordinates )
+{
+    DTK_REQUIRE( _global_ids.extent( 0 ) == _coordinates.extent( 0 ) );
```

Comment:
Wouldn't this not compile if they were not the same? If the device type is the same on both views per the type aliases then the memory and execution spaces must be the same as well